Source/Core/DolphinQt2/FIFOPlayerWindow.cpp

```diff
@@ -0,0 +1,354 @@
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOPlayerWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOPlayerWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!recorder.IsRecording())
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(), [this] {
+      QueueOnObject(&Settings::Instance(), [this] { OnRecordingDone(); });
+    });
+
+    m_save->setEnabled(false);
+    m_frame_range_to->setEnabled(false);
+    m_frame_range_from->setEnabled(false);
+    m_object_range_from->setEnabled(false);
+    m_object_range_to->setEnabled(false);
+    m_early_memory_updates->setEnabled(false);
+    m_frame_record_count->setEnabled(false);
+
+    m_record->setText(tr("Stop"));
+  }
+  else
+  {
+    recorder.StopRecording();
+  }
+
+  UpdateInfo();
+}
+
+void FIFOPlayerWindow::OnEmulationStarted()
+{
+  m_record->setEnabled(FifoPlayer::GetInstance().GetFile() == nullptr);
+  m_load->setEnabled(false);
+}
+
+void FIFOPlayerWindow::OnEmulationStopped()
+{
+  m_record->setEnabled(false);
+  m_load->setEnabled(true);
+
+  // If we have previously been recording, stop now.
+  if (FifoRecorder::GetInstance().IsRecording())
+    ToggleRecording();
+
+  if (FifoPlayer::GetInstance().GetFile() != nullptr && Core::IsRunning())
+  {
+    m_record->setEnabled(true);
+  }
+}
+
+void FIFOPlayerWindow::OnRecordingDone()
+{
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+  size_t fifo_bytes = 0;
+  size_t mem_bytes = 0;
+
+  for (u32 i = 0; i < file->GetFrameCount(); ++i)
+  {
+    fifo_bytes += file->GetFrame(i).fifoData.size();
+    for (const auto& mem_update : file->GetFrame(i).memoryUpdates)
+      mem_bytes += mem_update.data.size();
+  }
+
+  m_info_label->setText(tr("%1 FIFO bytes\n%2 memory bytes\n%3 frames")
+                            .arg(QString::number(fifo_bytes), QString::number(mem_bytes),
+                                 QString::number(file->GetFrameCount())));
+
+  m_save->setEnabled(true);
+  m_frame_range_to->setEnabled(true);
+  m_frame_range_from->setEnabled(true);
+  m_object_range_from->setEnabled(true);
+  m_object_range_to->setEnabled(true);
+  m_early_memory_updates->setEnabled(true);
+  m_frame_record_count->setEnabled(true);
+
+  m_record->setText(tr("Record"));
+}
+
+void FIFOPlayerWindow::UpdateInfo()
+{
+  if (FifoPlayer::GetInstance().GetFile() != nullptr && Core::IsRunning())
```

Comment:
This should be a method on FifoPlayer, like `bool FifoPlayer::IsPlaying()`. When it's in this class, it's 1) not very semantically clear (what does it mean the fifoplayer doesn't have a file?) and 2) more likely that changes to FifoPlayer can break this logic.

Reply:
Nope. I'll keep it as-is.

Reply:
Why not? Separation of concerns between presentation logic and other logic is a very core building block of UI programming, and results in more readable and less fragile code.

Reply:
Because this PR is supposed to implement a FIFO Player in Qt and *not* fix the shortcomings of the Core code it's using.

Reply:
If it were an architectural change or something with significant impact on other areas of code, I might agree, but in this case it's a couple added lines. I see no good reason not to put them in the correct place in this PR.